Source/WebCore/html/UserActivation.h

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright (C) 2022 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <wtf/IsoMalloc.h>
+#include <wtf/WeakPtr.h>
+
+namespace WebCore {
+
+class DOMWindow;
+class Navigator;
+
+class UserActivation final : public RefCounted<UserActivation> {
+    WTF_MAKE_ISO_ALLOCATED(UserActivation);
+
+public:
+    static Ref<UserActivation> create(Navigator&);
+    ~UserActivation();
+
+    Navigator* navigator();
```

Comment:
Given it seems from the tests MessageEvent might be able to own a UserActivation, is there a chance it gets exposed to workers as well?
I am not exactly sure how hasBeenActive/isActive would be computed in that case.
Maybe UserActivation would use its Navigator reference only in case it is called from a Document context?

In any case, I would tend to make navigator() private so that it is only used internally.

Reply:
I don't think we want to support the `MessageEvent` aspect (at least, not yet). It's not part of the HTML spec also and the PR this change is based on.  

Sorry, I should have made that clear in the (non-existent) commit message.